gmeta/codegen/src/lib.rs

```diff
@@ -1,518 +1,503 @@
+/// ```
+///
+/// # Syntax
+///
+/// - This attribute **must** be used on the `pub`lic `mod` container with the
+/// `metafuncs` identifier.
+/// - The first item in the module **must** be a `pub`lic `type` alias with the
+/// `State` identifier. The type for which `State` will be an alias **must**
+/// implement [`Decode`] trait.
+///
+/// Usually the state type should be imported from the implemented associative
+/// [`Metadata::State`](../gmeta/trait.Metadata.html#associatedtype.State) type
+/// from the contract's `io` crate.
+///
+/// - The rest of items **must** be `pub`lic functions.
+/// - The first argument's type of metafunctions **must** be `State`.
+///
+/// In addition to the mandatory first argument, functions can have additional
+/// ones.
+///
+/// - The maximum amount of additional arguments is 18 due restrictions of the
+/// SCALE codec.
+/// - All additional arguments **must** implement the [`Decode`] &
+/// [`TypeInfo`] traits.
+/// - A function **mustn't** return `()` or nothing.
+/// - A returned type **must** implement the
+/// [`Encode`](../gmeta/trait.Encode.html) & [`TypeInfo`] traits.
+///
+/// [`Decode`]: ../gmeta/trait.Decode.html
+/// [`TypeInfo`]: ../gmeta/trait.TypeInfo.html
+///
+/// # Expansion result
+///
+/// This attribute doesn't change the `metafuncs` module and items inside, but
+/// adds `use super::*;` inside the module because, in most cases, it'll be
+/// useful for importing items from an upper namespace. So every item in the
+/// same namespace where the module is located is accessible inside it.
+///
+/// The rest of the magic happens in the another generated private `extern`
+/// module. It registers all metawasm functions, their arguments & return types,
+/// and generates extern functions with the same names. Later, they can be
+/// called from a metaWASM binary inside a blockchain.
+///
+/// **Important note**: although metafunctions can take more than 1 additional
+/// arguments, on the metaWASM binary level, they must be passed as one. So if
+/// the amount of additinal arguments is 0 or 1, nothing needs to be changed,
+/// but if more - they all must be placed inside a tuple in the same order as in
+/// their function's signature.
+///
+/// E.g., argument definitions for the above example:
+/// - For `some_function` an argument must be [`None`].
+/// - For `another_function_but_with_arg` an argument must be `Some(SomeArg)`.
+/// - For `function_with_multiple_args` an argument must be
+/// `Some((SomeArg, u16, u32))`.
+#[proc_macro_attribute]
+pub fn metawasm(_: TokenStream, item: TokenStream) -> TokenStream {
+    let module: ItemMod = parse_macro_input!(item);
+    let module_span = module.span();
+
+    if let Err(error) = has_attributes(
+        module.attrs,
+        "module with #[metawasm] mustn't have attributes",
+    ) {
+        return error;
+    }
+
+    if let Err(error) = is_public(module_span, &module.vis) {
+        return error;
+    }
+
+    if module.ident != "metafuncs" {
```

Comment:
```suggestion
// somewhere above: const MOD_NAME: &str = "metafns";
    if module.ident != MOD_NAME {
```

Reply:
What's motivation?

Reply:
not to forget change other entries if renamed in future
here one occurrence in comparison and one in static str as error argument. reusing statics isn't the best way for me, but also not the deal, we can leave as it is